File: reader/DNAGroup.cpp
#include "DNAGroup.h"

DNAGroup::DNAGroup(string name) : m_name(name) {};
DNAGroup::~DNAGroup(void) {};

void DNAGroup::make_from_dgi(DatagramIterator &dgi, DNAStorage* /*store*/)
{
	string name, parent_name, vis_group_name;
	
	name = dgi_extract_string8(dgi);
	parent_name = dgi_extract_string8(dgi);
	vis_group_name = dgi_extract_string8(dgi);

	m_name = name;
};

DNAVisGroup* DNAGroup::get_vis_group()
{
	return m_vis_group;
};
		
void DNAGroup::add(DNAGroup* child)
{
	m_children.push_back(child);
};

DNAGroup* DNAGroup::at(size_t index)
{
	return m_children.at(index);
};

size_t DNAGroup::get_num_children()
{
	return m_children.size();
};
	
void DNAGroup::set_parent(DNAGroup* parent)
{
	m_parent = parent;
	m_vis_group = parent->get_vis_group();
	parent->add(this);
};

void DNAGroup::clear_parent()
{
	m_parent = NULL;
	m_vis_group = NULL;
};

DNAGroup* DNAGroup::get_parent()
{
	return m_parent;
};

string DNAGroup::get_name()
{
	return m_name;
};

void DNAGroup::traverse(NodePath& np, DNAStorage* store)
{
	NodePath _np = np.attach_new_node(m_name);
	for (comp_vec_t::iterator child = m_children.begin(); child != m_children.end(); child++)
	{
        (*child)->traverse(_np, store);
	};
};

File: libpandadna/DNAGroup.py
from common import *
from DNAError import DNAError


class DNAGroup:
    COMPONENT_CODE = 1

    def __init__(self, name):
        self.name = name
        self.children = []
        self.parent = None
        self.visGroup = None

    def setName(self, name):
        self.name = name

    def getName(self):
        return self.name

    def getCode(self):
        return self.code

    def add(self, child):
        self.children.append(child)

    def setVisGroup(self, visGroup):
        self.visGroup = visGroup

    def getVisGroup(self):
        return self.visGroup

    def setParent(self, parent):
        self.parent = parent
        self.visGroup = parent.getVisGroup()

    def getParent(self):
        return self.parent

    def clearParent(self):
        self.parent = None
        self.visGroup = None

    def at(self, index):
        return self.children[index]

    def getNumChildren(self):
        return len(self.children)

    def makeFromDGI(self, dgi):
        self.name = dgi.getString()
        dgi.getString()
        dgi.getString()

    def traverse(self, np, store):
        _np = np.attachNewNode(self.name)
        self.traverseChildren(_np, store)

    # convenience functions
    def raiseCodeNotFound(self, code=None):
        if code is None:
            code = self.code
        _error = '%s code (%s) not found in storage!'
        raise DNAError(_error % (self.__class__.__name__, code))

    def traverseChildren(self, np, store):
        for child in self.children:
            child.traverse(np, store)
